File: 2023/day-3/part_one.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"regexp"
	"strconv"
	"strings"
)

// A number might appear again in other lines.
// A special character might be touching multiple numbers with the same value.
// A special character might be touching the same number multiple times.

var NUMBERS = map[string]string{}

func main() {
	linesOfChars := Map(ReadLines("input.txt"), func(line string) []string {
		return strings.Split(line, "")
	})

	for x, line := range linesOfChars {
		for y, char := range line {
			r := regexp.MustCompile(`[^.\d]`)
			match := r.MatchString(char)

			if match {
				findAndAppendNumbers(linesOfChars, x, y)
			}
		}
	}

	sum := 0
	for _, number := range NUMBERS {
		n, _ := strconv.Atoi(number)
		sum += n
	}

	fmt.Println(sum)
	// 535078
}

func findAndAppendNumbers(chars [][]string, x int, y int) {
	xPositions := map[int]bool{
		x - 1: x > 0,
		x:     x >= 0,
		x + 1: x < len(chars)-1,
	}

	for xPosition, xCondition := range xPositions {
		if xCondition {
			yPositions := map[int]bool{
				y - 1: y > 0,
				y:     y >= 0,
				y + 1: y < len(chars[xPosition])-1,
			}

			for yPosition, yCondition := range yPositions {
				if yCondition {
					appendNumber(chars, xPosition, yPosition)
				}
			}
		}
	}
}

func appendNumber(chars [][]string, x int, y int) {
	char := chars[x][y]

	_, err := strconv.Atoi(char)
	if err != nil {
		return
	}

	digits := []string{char}

	i := 1
	keepGoing := true
	for keepGoing && y-i >= 0 {
		_, err := strconv.Atoi(chars[x][y-i])
		if err != nil {
			keepGoing = false
		} else {
			digits = append([]string{chars[x][y-i]}, digits...)
			i++
		}
	}

	j := 1
	keepGoing = true
	for keepGoing && y+j < len(chars[x]) {
		_, err := strconv.Atoi(chars[x][y+j])
		if err != nil {
			keepGoing = false
		} else {
			digits = append(digits, chars[x][y+j])
			j++
		}
	}

	position := fmt.Sprintf("%v%v%v", x, y-i+1, y+j-1) // For number with coords [58, 101->103] => 5810103
	NUMBERS[position] = strings.Join(digits, "")
}

func Map[T any, R any](sli []T, fun func(T) R) []R {
	res := make([]R, len(sli))
	for i, ele := range sli {
		res[i] = fun(ele)
	}

	return res
}

func ReadLines(fileName string) (lines []string) {
	dir, _ := os.Getwd()
	filePath := fmt.Sprintf("%v/%v", dir, fileName)
	file, _ := os.Open(filePath)

	defer file.Close()

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}

	return
}

File: 2023/day-3/part_one.ts
import fs from "fs"
import path from "path"

const NUMBERS: { [key: string]: string } = {}

const appendNumber = (chars: string[][], x: number, y: number) => {
  if (!/\d/.test(chars[x][y])) return

  const digits = [chars[x][y]]

  let i = 1
  let keepGoing = true
  while (keepGoing && y - i >= 0) {
    if (/\d/.test(chars[x][y - i])) {
      digits.unshift(chars[x][y - i])
      i++
    } else {
      keepGoing = false
    }
  }

  let j = 1
  keepGoing = true
  while (keepGoing && y + j < chars[x].length) {
    if (/\d/.test(chars[x][y + j])) {
      digits.push(chars[x][y + j])
      j++
    } else {
      keepGoing = false
    }
  }

  const position = `${x}${y - i - 1}${y + j - 1}`
  NUMBERS[position] = digits.join("")
}

const findAndAppendNumbers = (chars: string[][], x: number, y: number) => {
  const xPositions: [number, boolean][] = [[x - 1, x > 0], [x, x >= 0], [x + 1, x < chars.length - 1]]
  xPositions.forEach(([xPosition, xCondition]) => {
    if (xCondition) {
      const yPositions: [number, boolean][] = [[y - 1, y > 0], [y, y >= 0], [y + 1, y < chars.length - 1]]
      yPositions.forEach(([yPosition, yCondition]) => {
        if (yCondition) {
          appendNumber(chars, xPosition, yPosition)
        }
      })
    }
  })
}

fs.readFile(path.resolve(__dirname, "input.txt"), "utf8", (_, lines: string) => {
  const linesOfChars = lines.trim().split("\n").map(line => line.split(""))

  linesOfChars.forEach((line, x) => {
    line.forEach((char, y) => {
      if (/[^.\d]/.test(char)) {
        findAndAppendNumbers(linesOfChars, x, y)
      }
    })
  })

  const sum = Object.values(NUMBERS).reduce((sum, number) => sum + +number, 0)
  console.log(sum)
  // 535078
})
